src/core/server/elasticsearch/client/retry_unauthorized.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { MaybePromise } from '@kbn/utility-types';
+import { AuthHeaders, KibanaRequest, SetAuthHeaders, isRealRequest } from '../../http';
+import { ScopeableRequest } from '../types';
+import { UnauthorizedError } from './errors';
+
+export interface UnauthorizedErrorHandlerOptions {
+  error: UnauthorizedError;
+  request: KibanaRequest;
+}
+
+export type UnauthorizedErrorHandlerResultType = 'notHandled';
+
+export interface UnauthorizedErrorHandlerResultRetryParams {
+  authHeaders: AuthHeaders;
+}
+
+export interface UnauthorizedErrorHandlerRetryResult
+  extends UnauthorizedErrorHandlerResultRetryParams {
+  type: 'retry';
+}
+
+export interface UnauthorizedErrorHandlerNotHandledResult {
+  type: 'notHandled';
+}
+
+export type UnauthorizedErrorHandlerResult =
+  | UnauthorizedErrorHandlerRetryResult
+  | UnauthorizedErrorHandlerNotHandledResult;
+
+interface UnauthorizedErrorHandlerToolkit {
+  /**
+   * The handler cannot handle the error, or was not able to reauthenticate
+   * */
+  notHandled: () => UnauthorizedErrorHandlerResult;
+  /**
+   * The handle was able to reauthenticate.
+   * @param params
+   */
+  retry: (params: UnauthorizedErrorHandlerResultRetryParams) => UnauthorizedErrorHandlerResult;
+}
+
+export const toolkit: UnauthorizedErrorHandlerToolkit = {
+  notHandled: () => ({ type: 'notHandled' }),
+  retry: ({ authHeaders }) => ({
+    type: 'retry',
+    authHeaders,
+  }),
+};
+
+export type UnauthorizedErrorHandler = (
+  options: UnauthorizedErrorHandlerOptions,
+  toolkit: UnauthorizedErrorHandlerToolkit
+) => MaybePromise<UnauthorizedErrorHandlerResult>;
+
+export type InternalUnauthorizedErrorHandler = (
+  error: UnauthorizedError
+) => MaybePromise<UnauthorizedErrorHandlerResult>;
+
+/**
+ * Converts the public version of `UnauthorizedErrorHandler` to the internal one used by the ES client
+ *
+ * @internal
+ */
+export const createInternalErrorHandler = ({
+  handler,
+  request,
+  setAuthHeaders,
+}: {
+  handler: UnauthorizedErrorHandler;
+  request: ScopeableRequest;
+  setAuthHeaders: SetAuthHeaders;
+}): InternalUnauthorizedErrorHandler => {
+  // we don't want to support 401 retry for fake requests
+  if (!isRealRequest(request)) {
+    return () => toolkit.notHandled();
+  }
```

Comment:
Yeah, I _believe_ it should be enough, [here is the only place](https://github.com/elastic/kibana/blob/6693ef371f887eca639b09c4c9b15701b4ebabd4/x-pack/plugins/security/server/authentication/providers/base.ts#L122-L126) I'm aware of where we'd want to prevent automatic retry logic and we don't use real request there (`.asScoped({ headers: { ...request.headers, ...authHeaders } })`).